parsl/tests/test_radical/test_mpi_funcs.py

```diff
@@ -0,0 +1,23 @@
+import parsl
+import pytest
+
+from parsl.tests.configs.local_radical_mpi import fresh_config as local_config
+
+
+@parsl.python_app
+def test_mpi_func(comm, msg, sleep, ranks):
```

Comment:
So 2 options are here: 

1. We can set the `comm` as an argument but it must be the first one in the function definition/signature
2. We can set the `comm` as a keyword argument and it can be `None` anywhere.

This is how Radical Pilot identifies the `comm`

Reply:
I feel MPI based apps are going to have different semantics compared to regular python_apps (see also how bash_apps have different semantics to python_apps) - so I don't think we need to be too scared about adding in new parameters that only make sense in the context of a multi-rank MPI app.

Other bits of parsl use magic parameters in a few ways, without much consistency:

* `inputs` is a magic list that will get file *input* staging done on all of its members
* `outputs` is a magic list that will get file *output* staging done on all of its members
* `parsl_resource_specification` lets an app invoker make special requests for how it is executed (for example, amount of memory to budget when using the Work Queue Executor)

What little consistency there is suggests that `comm` should be a keyword argument, option 2 above, and that you don't need to specify it when invoking - instead it will be silently inserted by an MPI-aware executor.

`test_mpi_func(msg='mpi.func.%06d' % i, sleep=1, ranks=2)`

rather than 

` test_mpi_func(None, msg='mpi.func.%06d' % i, sleep=1, ranks=2, comm=None)`

I'm not sure how this overlaps with ideas that Yadu has had about multirank python apps: so far in his prototype (#2905 ) we mostly talked about command-line based stuff which looks, I think, quite different.

Reply:
That makes sense. If the executor passes in a communicator, and the function makes use of it, then it couldn't possibly run on any of the other executors currently supported. A new `mpi_python_app` that's separate from `python_app` might be more appropriate. This would be required since a single invocation of this app with ranks>1 is expected to launch multiple instances of the function, and we'd need to think through existing options like `inputs`, `outputs`, and `stdout` in this context. 

For now, setting `comm` to be a special kwargs should be good enough to be consistent. 

For now, to get this feature across the finish line,

Reply:
This is now addressed here: https://github.com/Parsl/parsl/pull/2923/commits/0e4c709fb833650bc74b78f0ae00409fe56f5745